File: Top200/377-CombinationSum4.go
package main

func combinationSum4(nums []int, target int) int {
	l := len(nums)

	if l == 0 {
		return 0
	}

	dp := make([]int, target+1)
	dp[0] = 1

	for i := 1; i <= target; i++ {
		for j := 0; j < l; j++ {
			if nums[j] <= i {
				dp[i] += dp[i-nums[j]]
			}
		}
	}

	return dp[target]
}

File: Top200/377-CombinationSum4.java
package Top200;

class Solution {
    public int combinationSum4(int[] nums, int target) {
        // dp 数组定义
        int[] dp = new int[target + 1];

        // dp 数组初始化
        dp[0] = 1;

        // dp 迭代
        for (int i = 1; i <= target; i++) {
            for (int j = 0; j < nums.length; j++) {
                if (i >= nums[j]) {
                    dp[i] += dp[i-nums[j]];
                }
            }
        }

        return dp[target];
    }
}
